node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+//! Generates rust code from the proto files.
+//!
+//! Protobuf files are collected recursively from $CARGO_MANIFEST_DIR/<input_root>/ directory.
+//! Corresponding "cargo:rerun-if-changed=..." line is printed to stdout, so that
+//! the build script running this function is rerun whenever proto files change.
+//! A single rust file is generated and stored at $OUT_DIR/<input_root>/gen.rs file.
+//!
+//! Protobuf files are compiled to a protobuf descriptor stored at
+//! $OUT_DIR/<input_root>/gen.binpb.
+//! Additionally a "PROTOBUF_DESCRIPTOR=<absolute path to that descriptor>" line is printed to
+//! stdout. This can be used to collect all the descriptors across the build as follows:
+//! 1. Checkout the repo to a fresh directory and then run "cargo build --all-targets"
+//!    We need it fresh so that every crate containing protobufs has only one build in the
+//!    cargo cache.
+//! 2. grep through all target/debug/build/*/output files to find all "PROTOBUF_DESCRIPTOR=..."
+//!    lines and merge the descriptor files by simply concatenating them.
+//! Note that you can run this procedure for 2 revisions of the repo and look for breaking
+//! changes by running "buf breaking <after.binpb> --against <before.binpb>" where before.binpb
+//! and after.binpb are the concatenated descriptors from those 2 revisions.
+//!
+//! The proto files are not expected to be self-contained - to import proto files from
+//! different crates you need to specify them as dependencies in the Config.dependencies.
+//! It is not possible to depend on a different proto bundle within the same crate (because
+//! these are being built simultaneously from the same build script).
+//!
+//! cargo build --all-targets
+//! perl -ne 'print "$1\n" if /PROTOBUF_DESCRIPTOR="(.*)"/' `find target/debug/build/*/output -type f` | xargs cat > /tmp/sum.binpb
+//! buf breaking /tmp/sum.binpb --against /tmp/sum.binpb
+#![allow(clippy::print_stdout)]
+use anyhow::Context as _;
+pub use once_cell::sync::Lazy;
+// Imports accessed from the generated code.
+pub use prost;
+use prost::Message as _;
+pub use prost_reflect;
+use std::{fs, path::Path, sync::Mutex};
+pub use syntax::*;
+
+mod canonical;
+mod ident;
+mod syntax;
+
+/// Traversed all the files in a directory recursively.
+fn traverse_files(
+    path: &Path,
+    f: &mut impl FnMut(&Path) -> anyhow::Result<()>,
+) -> anyhow::Result<()> {
+    if !path.is_dir() {
+        f(path).with_context(|| path.to_str().unwrap().to_string())?;
```

Comment:
Nit: `Path` has a [`display()`](https://doc.rust-lang.org/std/path/struct.Path.html#method.display) method to conveniently and infallibly convert it to a displayable string